### PR TITLE
Startup

### PR DESCRIPTION
**Pull request by iGxnon**
Please briefly answer these questions:

Base on #42 

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

* what changes does this pull request make?

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

### CODE REVIEW

File: api/v1alpha1/xlinecluster_types.go [hunk 1]
```diff
@@ -58,6 +58,9 @@ type XlineClusterSpec struct {
 	// Xline cluster image
 	Image *string `json:"image,omitempty"`
 
+	/// Xline container arguments
+	Args []string `json:"args,omitempty"`
```
Comment (Phoenix500526) [inline api/v1alpha1/xlinecluster_types.go:62]: IMO, define a `BootupArgs` is a better choice in that the kubebuilder can help us validate the field name and its type.

File: internal/transformer/xlinecluster_resource.go [hunk 2]
```diff
@@ -83,8 +83,20 @@ func MakeStatefulSet(cr *xapi.XlineCluster, scheme *runtime.Scheme) *appv1.State
 		Ports: []corev1.ContainerPort{
 			{Name: "xline-port", ContainerPort: 2379},
 		},
+		Command: []string{
```
Comment (Phoenix500526) [inline internal/transformer/xlinecluster_resource.go:86]: We should use the args defined in XlineCluster to cover the default bootup command.

File: api/v1alpha1/zz_generated.deepcopy.go [hunk 3]
```diff
@@ -128,6 +128,11 @@ func (in *XlineClusterSpec) DeepCopyInto(out *XlineClusterSpec) {
 		*out = new(string)
 		**out = **in
 	}
+	if in.Args != nil {
+		in, out := &in.Args, &out.Args
+		*out = make([]string, len(*in))
+		copy(*out, *in)
```
Comment (themanforfree) [inline api/v1alpha1/zz_generated.deepcopy.go:134]: Can it be changed to this? Why do we need two layers of pointers here?
```go
in, out := in.Args, out.Args
out = make([]string, len(in))
copy(out, in)
```

Comment (iGxnon) [reply api/v1alpha1/zz_generated.deepcopy.go:134]: This file is generated, I cannot understand it either.
But it works.

File: internal/transformer/xlinecluster_resource.go (hunk 2 above)
Comment (iGxnon) [reply internal/transformer/xlinecluster_resource.go]: done

**Review by themanforfree — DISMISSED** [review]
LGTM

File: api/v1alpha1/xlinecluster_types.go [hunk 4]
```diff
@@ -47,6 +49,50 @@ type XlineClusterList struct {
...
+	AutoRevisionRetention       *string `json:"autoRevisionRetention,omitempty"`
+	InitialClusterState         *string `json:"initialClusterState,omitempty"`
+	RetryCount                  *int    `json:"retryCount,omitempty"`
+	BatchMaxSize                *int    `json:"batchMaxSize,omitempty"`
+	FollowerTimeoutTicks        *int    `json:"followerTimeoutTicks,omitempty"`
+	CandidateTimeoutTicks       *int    `json:"candidateTimeoutTicks,omitempty"`
+	LogEntriesCap               *int    `json:"logEntriesCap,omitempty"`
+	CmdWorkers                  *int    `json:"cmdWorkers,omitempty"`
+	CompactBatchSize            *int    `json:"compactBatchSize,omitempty"`
+	Quota                       *int    `json:"quota,omitempty"`
```
Comment (Phoenix500526) [inline api/v1alpha1/xlinecluster_types.go:93]: please modify the json field name in the annotation, for example,  modify `json:"jaegerOffline,omitempty"` to `json:"jaeger-offline,omitempty"`.

File: api/v1alpha1/xlinecluster_types.go [hunk 5]
```diff
@@ -66,6 +116,129 @@ type XlineClusterSpec struct {
...
+		args = append(args, "--cmd-workers", strconv.Itoa(*s.BootstrapArgs.CmdWorkers))
+	}
+	if s.BootstrapArgs.CompactBatchSize != nil {
+		args = append(args, "--compact-batch-size", strconv.Itoa(*s.BootstrapArgs.CompactBatchSize))
+	}
+	if s.BootstrapArgs.Quota != nil {
+		args = append(args, "--quota", strconv.Itoa(*s.BootstrapArgs.Quota))
+	}
+	return args
+}
```
Comment (Phoenix500526) [inline api/v1alpha1/xlinecluster_types.go:240]: We can directly use `json.Marshal` to serialize the XlineArgs and then convert it to string. Any empty field will be skipped. We don't need to check every field in the XlineArgs one by one.

File: api/v1alpha1/xlinecluster_types.go [hunk 6]
```diff
@@ -47,6 +49,50 @@ type XlineClusterList struct {
...
+	RangeRetryTimeout           *string `json:"rangeRetryTimeout,omitempty"`
+	CompactTimeout              *string `json:"compactTimeout,omitempty"`
+	SyncVictimsInterval         *string `json:"syncVictimsInterval,omitempty"`
+	WatchProgressNotifyInterval *string `json:"watchProgressNotifyInterval,omitempty"`
+	CurpDir                     *string `json:"curpDir,omitempty"`
+	CompactSleepInterval        *string `json:"compactSleepInterval,omitempty"`
+	AutoCompactMode             *string `json:"autoCompactMode,omitempty"`
+	AutoPeriodicRetention       *string `json:"autoPeriodicRetention,omitempty"`
+	AutoRevisionRetention       *string `json:"autoRevisionRetention,omitempty"`
+	InitialClusterState         *string `json:"initialClusterState,omitempty"`
```
Comment (Phoenix500526) [inline api/v1alpha1/xlinecluster_types.go:85]: Same as `isLeader`

File: api/v1alpha1/xlinecluster_types.go (hunk 4 above)
Comment (iGxnon) [reply api/v1alpha1/xlinecluster_types.go:93]: Need to maintain the same specification as k8s yaml.

File: api/v1alpha1/xlinecluster_types.go [hunk 7]
```diff
@@ -47,6 +49,50 @@ type XlineClusterList struct {
 	Items           []XlineCluster `json:"items"`
 }
 
+// XlineArgs
+// +k8s:openapi-gen=true
```
[the remark this answers is not in the document]
Comment (iGxnon) [reply api/v1alpha1/xlinecluster_types.go:53]: Otherwise, `CR` cannot be verified in the k8s API.

File: api/v1alpha1/xlinecluster_types.go (hunk 5 above)
Comment (iGxnon) [reply api/v1alpha1/xlinecluster_types.go:240]: json format differs from arguments list much. When parsing a JSON string, writing some fragile code, such as trimming idents and separators.

File: api/v1alpha1/xlinecluster_types.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (Phoenix500526) [reply api/v1alpha1/xlinecluster_types.go:53]: The `XlineArgs` is a sub-resource of the `XlineClusterSpec`, and the verification of the latter will cover the `XlineArgs`

File: api/v1alpha1/xlinecluster_types.go (hunk 5 above)
Comment (Phoenix500526) [reply api/v1alpha1/xlinecluster_types.go:240]: Here is an example:
```go
```

File: api/v1alpha1/xlinecluster_types.go (hunk 4 above)
Comment (Phoenix500526) [reply api/v1alpha1/xlinecluster_types.go:93]: The `bootstrapArgs` meets the specification of k8s yaml

**Comment by Phoenix500526** [comment]
Please make sure that the e2e test cover the modification of this pr.

File: api/v1alpha1/xlinecluster_types.go [hunk 8]
```diff
@@ -47,6 +49,50 @@ type XlineClusterList struct {
...
+	SyncVictimsInterval         *string `json:"syncVictimsInterval,omitempty"`
+	WatchProgressNotifyInterval *string `json:"watchProgressNotifyInterval,omitempty"`
+	CurpDir                     *string `json:"curpDir,omitempty"`
+	CompactSleepInterval        *string `json:"compactSleepInterval,omitempty"`
+	AutoCompactMode             *string `json:"autoCompactMode,omitempty"`
+	AutoPeriodicRetention       *string `json:"autoPeriodicRetention,omitempty"`
+	AutoRevisionRetention       *string `json:"autoRevisionRetention,omitempty"`
+	InitialClusterState         *string `json:"initialClusterState,omitempty"`
+	RetryCount                  *int    `json:"retryCount,omitempty"`
+	BatchMaxSize                *int    `json:"batchMaxSize,omitempty"`
```
Comment (Phoenix500526) [inline api/v1alpha1/xlinecluster_types.go:87]: Why `*int` not `int`?

Comment (iGxnon) [reply api/v1alpha1/xlinecluster_types.go]: done

File: api/v1alpha1/xlinecluster_types.go [hunk 9]
```diff
@@ -47,6 +49,43 @@ type XlineClusterList struct {
...
+	ClientMaxRetryTimeout       *string `json:"client-max-retry-timeout,omitempty"`
+	GcInterval                  *string `json:"gc-interval,omitempty"`
+	RangeRetryTimeout           *string `json:"range-retry-timeout,omitempty"`
+	CompactTimeout              *string `json:"compact-timeout,omitempty"`
+	SyncVictimsInterval         *string `json:"sync-victims-interval,omitempty"`
+	WatchProgressNotifyInterval *string `json:"watch-progress-notify-interval,omitempty"`
+	CurpDir                     *string `json:"curp-dir,omitempty"`
+	CompactSleepInterval        *string `json:"compact-sleep-interval,omitempty"`
+	RetryCount                  int     `json:"retry-count,omitempty"`
+	BatchMaxSize                int     `json:"batch-max-size,omitempty"`
```
Comment (Phoenix500526) [inline api/v1alpha1/xlinecluster_types.go:80]: The `batch-max-size` should be string.

File: api/v1alpha1/xlinecluster_types.go [hunk 10]
```diff
@@ -47,6 +49,43 @@ type XlineClusterList struct {
...
+	JaegerOffline               bool    `json:"jaeger-offline,omitempty"`
+	JaegerOnline                bool    `json:"jaeger-online,omitempty"`
+	JaegerLevel                 bool    `json:"jaeger-level,omitempty"`
+	ClientUseBackoff            bool    `json:"client-use-backoff,omitempty"`
+	AuthPrivateKey              *string `json:"auth-private-key,omitempty"`
+	AuthPublicKey               *string `json:"auth-public-key,omitempty"`
+	JaegerOutputDir             *string `json:"jaeger-output-dir,omitempty"`
+	LogFile                     *string `json:"log-file,omitempty"`
+	LogRotate                   *string `json:"log-rotate,omitempty"`
+	LogLevel                    *string `json:"log-level,omitempty"`
```
Comment (Phoenix500526) [inline api/v1alpha1/xlinecluster_types.go:62]: It's better to use `kubebuilder:validation:Enum` to specify all the possible values for enum args.  FYI: https://book.kubebuilder.io/reference/markers/crd-validation

File: api/v1alpha1/xlinecluster_types.go [hunk 11]
```diff
@@ -47,6 +49,43 @@ type XlineClusterList struct {
...
+	BatchTimeout                *string `json:"batch-timeout,omitempty"`
+	ClientWaitSyncedTimeout     *string `json:"client-wait-synced-timeout,omitempty"`
+	ClientProposeTimeout        *string `json:"client-propose-timeout,omitempty"`
+	ClientInitialRetryTimeout   *string `json:"client-initial-retry-timeout,omitempty"`
+	ClientMaxRetryTimeout       *string `json:"client-max-retry-timeout,omitempty"`
+	GcInterval                  *string `json:"gc-interval,omitempty"`
+	RangeRetryTimeout           *string `json:"range-retry-timeout,omitempty"`
+	CompactTimeout              *string `json:"compact-timeout,omitempty"`
+	SyncVictimsInterval         *string `json:"sync-victims-interval,omitempty"`
+	WatchProgressNotifyInterval *string `json:"watch-progress-notify-interval,omitempty"`
```
Comment (Phoenix500526) [inline api/v1alpha1/xlinecluster_types.go:76]: It's better to use `kubebuilder:validation:Pattern` to specify all the interval args.

File: api/v1alpha1/xlinecluster_types.go [hunk 12]
```diff
@@ -47,6 +49,43 @@ type XlineClusterList struct {
...
+	CurpDir                     *string `json:"curp-dir,omitempty"`
+	CompactSleepInterval        *string `json:"compact-sleep-interval,omitempty"`
```
Comment (Phoenix500526) [inline api/v1alpha1/xlinecluster_types.go:78]: Same as above

File: api/v1alpha1/xlinecluster_types.go [hunk 13]
```diff
@@ -47,6 +49,43 @@ type XlineClusterList struct {
...
+	RetryCount                  int     `json:"retry-count,omitempty"`
+	BatchMaxSize                int     `json:"batch-max-size,omitempty"`
+	FollowerTimeoutTicks        int     `json:"follower-timeout-ticks,omitempty"`
+	CandidateTimeoutTicks       int     `json:"candidate-timeout-ticks,omitempty"`
+	LogEntriesCap               int     `json:"log-entries-cap,omitempty"`
+	CmdWorkers                  int     `json:"cmd-workers,omitempty"`
```
Comment (Phoenix500526) [inline api/v1alpha1/xlinecluster_types.go:84]: It's better to use `kubebuilder:validation:Minimum` to make sure all these numeric args are larger than or equal to 1.

**Review by Phoenix500526 — APPROVED** [review]
LGTM

**Review by themanforfree — APPROVED** [review]

**Pull request merged**